game-app/game-core/src/main/java/games/strategy/triplea/ui/UnitChooser.java

```diff
@@ -648,12 +659,149 @@ public Dimension getMinimumSize() {
+   *
+   * (With <code>ChooserEntry</code> being a candidate to go into the model
+   * and <code>UnitChooserEntryIcon</code> being a candidate to go into the presentation,
+   * the method <code>getImage</code> should probably move to <code>UnitChooserEntryIcon</code>.
+   *
+   * It can't be there for the moment, because <code>UnitChooserEntryIcon</code> is an
+   * inner class, and static methods are not supported in inner classes at language level 11.)
+   *
+   * @param category category of the image
+   * @param forceDamaged true iff the image should show a damaged unit
+   * @param uiContext context for loading the image resource
+   * @return the unit image, possibly decorated by the non-withdrawable-icon.
+   */
+
+  @VisibleForTesting
+  public static Image getImage(final UnitCategory category,
+                               final boolean forceDamaged,
+                               final UiContext uiContext) {
+    final var uiContextDecoration = UiContextDecoration.get(uiContext);
+    final Image image = uiContextDecoration.unitImageFactory
+        .getImage(
+            ImageKey.builder()
+                .type(category.getType())
+                .player(category.getOwner())
+                .damaged(forceDamaged || category.hasDamageOrBombingUnitDamage())
+                .disabled(category.getDisabled())
+                .build())
+        .orElseThrow();
+
+    return category.getCanRetreat()
+        ? image
+        : uiContextDecoration.getUnitImageWithNonWithdrawableImage(image);
+  }
+
+  /**
+   * If it would produce less code (e.g. using Kotlin), <code>UiContextDecoration</code>
+   * would be a proper application if the decorator pattern.
+   */
+
+  static class UiContextDecoration {
+    private static UiContextDecoration latestUCD;
+
+    @VisibleForTesting
+    public final UiContext uiContext;
+    @VisibleForTesting
+    public final UnitImageFactory unitImageFactory;
+    private final BufferedImage nonWithdrawableImage;
+    private final Map<Image,BufferedImage> unitImageWithNonWithrawableImage =
+        Collections.synchronizedMap(new HashMap<>());
+
+    private UiContextDecoration(final UiContext uiContext) {
+      this.uiContext = uiContext;
+      final int imageHeightInGame = uiContext.getUnitImageFactory().getUnitImageHeight();
+      Image img = uiContext.loadImage("misc/non-withdrawable_small.png").orElseThrow();
+      double scaleChange = 1.0;
+      if (imageHeightInGame < 2*img.getHeight(null)) {
+        scaleChange = 2.0*img.getHeight(null) / imageHeightInGame;
+      } else {
+        final Image imageBig = uiContext.loadImage("misc/non-withdrawable.png").orElseThrow();
+        if (imageHeightInGame >= 2*imageBig.getHeight(null)) {
+          img = imageBig;
+          if (imageHeightInGame >= 4*img.getHeight(null)) {
+            scaleChange = 2.0*img.getHeight(null) / imageHeightInGame;
+          }
+        }
+      }
+
```

Comment:
Reworking the code I realized that ```if (imageHeightInGame < 2*img.getHeight(null))``` is never true with the given data: The small image is 24px high, and the smallest unit icons I know are 48px high.
What is your coding philisophy: Should I remove the code for a situation that never happens in reality? (If I keep the code, I'd need to mockup testdata for that situation, which seems like a wate of effort to me. - Yes, I did not thinks about unit tests for this part of the code, but I now think that unit tests are necessary for this, not matter how much I can clear up the coding.)

Reply:
RE: philosophy on this

The most important is the advertised API of a method should work. There should be no way that given any input it'll produce an incorrect result or unexpectedly error out. If there are any such pre-conditions, they are best avoided (as it's a source of brittleness and cmoplexity), or otherwise should be enforced in one way or another.

In this case enforcing image sizes would happen at the time of map submission. That seems impractical.

It's still difficult for me to really say what the end goal of this code block is. Could you elaborate on that?

Reply:
Will be completely rewritten in the next commit  - including a correction, so thanks for pointing towards it.